### PR TITLE
ceph: bash auto complete for CLI based on mon command descriptions

### PR DESCRIPTION
**Pull request by aclamk**
This is not bind to bash yet. Use as 'ceph --comp osd ls'.
Able to fulfill commands as well as print command line help, when matching is complete.
This is required step before adding hinting of string values, eg. pool names.
Signed-off-by: Adam Kupczyk(a.kupczyk@mirantis.com)

### CODE REVIEW

**Comment by tchaikov** [comment]
@aclamk Adam, i assume that this pr supersedes #6264, am i right?

**Comment by aclamk** [comment]
@tchaikov Yes, #6264 is no longer useful.

File: src/ceph.in [hunk 1]
```diff
@@ -486,46 +486,40 @@ def complete(sigdict, args, target):
...
+                #no elements that match
+                matches = False
+                break
+            if not sig[j].N:
+                j = j + 1
+
+        if not matches:
+            continue
+
+        if j < len(sig) and len(args)>0:
```
Comment (tchaikov) [inline src/ceph.in]: @aclamk could you run this script through pep8? i see some of your change does not follow it.

File: src/ceph.in [hunk 2]
```diff
@@ -486,46 +486,40 @@ def complete(sigdict, args, target):
...
-            else:
-                d = sig[partindex]
-            comps.append(str(d))
-    if complete_verbose:
-        print >> sys.stderr, '\n'.join(comps)
-    print '\n'.join(comps)
+        i = 0
+        j = 0
+        matches = True
+        for i in range(0,len(args)-1):
```
Comment (tchaikov) [inline src/ceph.in]: could also put
```py
matches = False
j = 0
for arg in args:
  if j > len(sig) - 1:
    break
  arg_type = sig[j]
  matches = arg in arg_type.complete(arg)
  if not matches and arg_type.req:
    break
  if not arg_type.N:
    j += 1
else:
  matches = True
```

**Comment by tchaikov** [comment]
@aclamk and you might need to revise your commit message with more concise title and a `Signed-off-by` line, see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#signing-contributions

File: src/ceph.in (hunk 1 above)
Comment (aclamk) [inline src/ceph.in]: I will.

File: src/ceph.in (hunk 2 above)
Comment (aclamk) [inline src/ceph.in]: I will check analyze it first thing in the morning. Thank you for hints. This is actually my very thirst code in python.

File: src/ceph.in (hunk 1 above)
Comment (tchaikov) [inline src/ceph.in]: okay. fwiw, you can install it using `pip install pep8`.

**Comment by liewegas** [comment]
@aclamk do you mind rebasing this on master to resolve the conflict?  thanks!

**Comment by liewegas** [comment]
2016-03-22T07:24:07.627 INFO:teuthology.orchestra.run.smithi024.stderr:  File "/bin/ceph", line 523
2016-03-22T07:24:07.627 INFO:teuthology.orchestra.run.smithi024.stderr:    print '\n'.join(sorted(set(comps)))
2016-03-22T07:24:07.628 INFO:teuthology.orchestra.run.smithi024.stderr:             ^
2016-03-22T07:24:07.628 INFO:teuthology.orchestra.run.smithi024.stderr:SyntaxError: invalid syntax

File: src/ceph.in [hunk 3]
```diff
@@ -491,46 +492,38 @@ def complete(sigdict, args, target):
...
+            if j < len(sig) and len(args) > 0:
+                comps = comps + sig[j].complete(args[-1])
+
+            match_count = match_count + 1
+            match_cmd = cmd
 
+    if match_count == 1 and len(comps) == 0:
+        # only one command matched and no hints yet => add help
+        comps = comps + [' ', '#'+match_cmd['help']]
+    print '\n'.join(sorted(set(comps)))
```
Comment (tchaikov) [inline src/ceph.in]: @aclamk might need to fix this.

File: src/pybind/ceph_argparse.py [hunk 4]
```diff
@@ -219,6 +222,12 @@ def __str__(self):
             b += '(goodchars {0})'.format(self.goodchars)
         return '<string{0}>'.format(b)
 
+    def complete(self, s):
+        if s=='':
```
Comment (tchaikov) [inline src/pybind/ceph_argparse.py]: add a space before and after `==`.

File: src/ceph.in [hunk 5]
```diff
@@ -491,46 +492,38 @@ def complete(sigdict, args, target):
...
+            # successfully matched all - except last one - arguments
+            if j < len(sig) and len(args) > 0:
+                comps = comps + sig[j].complete(args[-1])
+
+            match_count = match_count + 1
+            match_cmd = cmd
 
+    if match_count == 1 and len(comps) == 0:
+        # only one command matched and no hints yet => add help
+        comps = comps + [' ', '#'+match_cmd['help']]
```
Comment (tchaikov) [inline src/ceph.in]: could you add space before and after `+`? guess pep8 will catch it..

File: src/ceph.in [hunk 6]
```diff
@@ -491,46 +492,38 @@ def complete(sigdict, args, target):
...
-            else:
-                d = sig[partindex]
-            comps.append(str(d))
-    if complete_verbose:
-        print('\n'.join(comps), file=sys.stderr)
-    print('\n'.join(comps))
+        j = 0
+        # iterate over all arguments, except last one
+        for arg in args[0:-1]:
+            if j > len(sig)-1:
```
Comment (tchaikov) [inline src/ceph.in]: could you please add spaces around `-`?

File: src/ceph.in [hunk 7]
```diff
@@ -491,46 +492,38 @@ def complete(sigdict, args, target):
...
+            found_match = arg in sig[j].complete(arg)
+            if not found_match and sig[j].req:
+                # no elements that match
+                break
+            if not sig[j].N:
+                j += 1
+        else:
+            # successfully matched all - except last one - arguments
+            if j < len(sig) and len(args) > 0:
+                comps = comps + sig[j].complete(args[-1])
```
Comment (tchaikov) [inline src/ceph.in]: could also put
```py
comps += sig[j].complete(args[-1])
```

File: src/ceph.in [hunk 8]
```diff
@@ -491,46 +492,38 @@ def complete(sigdict, args, target):
...
+
+            match_count = match_count + 1
```
Comment (tchaikov) [inline src/ceph.in]: could also put
```py
match_count += 1
```

**Comment by tchaikov** [comment]
@aclamk could you please add the bind to bash in "src/bash_completion/ceph", so we can include this feature in jewel?

File: src/ceph.in (hunk 8 above)
Comment (aclamk) [inline src/ceph.in]: 1) thanks for python tips. applied.
2) src/bash_completion/ceph added.

File: src/bash_completion/ceph [hunk 9]
```diff
@@ -1,73 +1,41 @@
 #
 # Ceph - scalable distributed file system
 #
-# Copyright (C) 2011 Wido den Hollander <wido@widodh.nl>
-#
 # This is free software; you can redistribute it and/or
 # modify it under the terms of the GNU Lesser General Public
 # License version 2.1, as published by the Free Software
-# Foundation.  See file COPYING.
+# Foundation.
```
Comment (tchaikov) [inline src/bash_completion/ceph]: trailing space.

**Comment by tchaikov** [comment]
@aclamk could you add your "Signed-off-by" to your commit of 4bffbd3? or you could just squash it into the previous one.

File: src/ceph.in [hunk 10]
```diff
@@ -512,9 +512,9 @@ def complete(sigdict, args, target):
         else:
             # successfully matched all - except last one - arguments
             if j < len(sig) and len(args) > 0:
-                comps = comps + sig[j].complete(args[-1])
+                comps += sig[j].complete(args[-1])
```
Comment (tchaikov) [inline src/ceph.in]: better off moving this change to the previous commit.

File: src/ceph.in [hunk 11]
```diff
@@ -512,9 +512,9 @@ def complete(sigdict, args, target):
...
 
-            match_count = match_count + 1
+            match_count += 1
```
Comment (tchaikov) [inline src/ceph.in]: ditto

File: src/bash_completion/ceph [hunk 12]
```diff
@@ -1,73 +1,41 @@
...
+	COMPREPLY=( $(compgen -W "${options_noarg} ${options_arg}" -- $cur) )
+	return 0
+    fi
+    declare -A hint_args
+    for (( i=1 ; i<cnt ; i++ ))
+    do
+	#skip this word if it is option
+	if [[ " ${options_noarg} " =~ " ${COMP_WORDS[i]} " ]] ;	then continue; fi
+	#skip this word and next if it is option with arg
+	if [[ " ${options_arg} " =~ " ${COMP_WORDS[i]} " ]] ;   then ((i++)); continue;	fi
```
Comment (tchaikov) [inline src/bash_completion/ceph]: could remove the extra spaces? or break the `if` block into multiple lines.

**Comment by tchaikov** [comment]
tested, works great! 

lgtm apart from some formatting nits.

**Comment by aclamk** [comment]
@tchaikov: added signed-off, reformatted bash code, removed spaces. Please take a look and comment.

**Comment by tchaikov** [comment]
@aclamk could you remove the heading spaces in the commit message of f9591f5
```
    Signed-off-by: Adam Kupczyk <a.kupczyk@mirantis.com>
```

and add a prefix of "bash_completion: " to the first line of it. and the commit message is not accurate, imho. we have a bash-completion for ceph CLI, but your change improves it by using the updated "ceph --completion".

**Comment by aclamk** [comment]
@tchaikov: Its a pleasure to polish out work with you.

**Comment by tchaikov** [comment]
lgtm

**Comment by tchaikov** [comment]
my pleasure also! =)

File: src/bash_completion/ceph [hunk 13]
```diff
@@ -1,73 +1,50 @@
 #
 # Ceph - scalable distributed file system
 #
-# Copyright (C) 2011 Wido den Hollander <wido@widodh.nl>
```
Comment (ErwanAliasr1) [inline src/bash_completion/ceph]: Why do you delete that copyright ?

**Comment by aclamk** [comment]
@ErwanAliasr1 
1) I deleted copyright intentionally
2) I was thinking that after change of this magnitude (actually all content new) it would be unfair to leave original copyright
3) I didnt put my copyright because I felt more along copyleft

**Comment by ErwanAliasr1** [comment]
@aclamk ok make sense.

**Pull request merged**